### ISSUE TITLE
Pre-run cells for intensive computation

### ISSUE THREAD

**Issue by willsimmons1465** [jupyter/jupyter-sphinx#140]
I am currently adding jupyter-sphinx to the docs for a project where some methods are especially expensive to run (e.g. connecting to a slow cloud service for running a job) or not feasible to run on our CI runners (e.g. not wanting to use private API keys and credits for the cloud service just for CI and docs building), and would prefer being able to just run it once locally and use the output.

I could not see such this feature available from looking around the docs or looking through the scripts in this repo, but if this is already possible that would be great. This is already available in nbsphinx, but I would prefer to see this in jupyter-sphinx for the added benefit of working inline from the .rst files.

Feature requested: insert the output from locally running the contents of the cell into the .rst file and have it display as if the cell had been run, using the same formatting as usual.

Possible implementation: a `:dont-run:` directive to skip the execution of the cell and just pass it straight to the cell formatting

Temporary hack idea: the first with the code using `:hide-output:` and `:raises:` to accept that it will throw an error when there is no API key found, and the second with just `print(...)` and `:hide-code:` to simulate the output. This would still mean the expensive calls will be made when I am just trying to build the docs locally, which kinda sucks.

**Comment by akhmerov** [comment]
Thanks for the suggestion!

I'd like to understand the use case and the possible implementation better. From your description, it seems that you want a part of the documentation to be computable, but not when building documentation.

If you were using current tools, you'd need to store this data somewhere (likely in the repository), and then use either of the two options:

- Write a jupyter-sphinx cell that only loads and renders the output
- Actually include the output in the .rst using standard tools

What are the disadvantages of these approaches, and how would an alternative look like?

**Comment by willsimmons1465** [comment]
> I'd like to understand the use case and the possible implementation better. From your description, it seems that you want a part of the documentation to be computable, but not when building documentation.

Yes. The particular operations in question are accessing a public access quantum computer to run jobs on it, where supply of credits for the quantum computer are limited and expensive, and access is typically shared with other users (I don't particularly want our regular docs builds to clog up the job queues on these machines and slow down actual important research use). So any solution that would allow us to demonstrate how to submit these jobs but doesn't actually run them during docs builds is good.

> If you were using current tools, you'd need to store this data somewhere (likely in the repository), and then use either of the two options:
> 
> * Write a jupyter-sphinx cell that only loads and renders the output
> * Actually include the output in the .rst using standard tools
> 
> What are the disadvantages of these approaches, and how would an alternative look like?

Either of these would work in the sense of presenting the right information, but I personally do not know how to get the formatting and style consistent with jupyter-sphinx cells using the standard rst tools (we would still need to do this for the code sample using either of the approaches you suggest). If you have an easy way to set up the styling of a sphinx project such that the standard code blocks are rendered using the same style as jupyter-sphinx cells, then please let me know (this might be something useful to add to the jupyter-sphinx documentation).

For the sake of example, suppose the `MyExpensiveCall()` operation would take 15 minutes to execute in practice (i.e. much longer than I would like to spend building the entire docs), and gives the following output when run in a jupyter cell:

```
Connecting... done
Job submitted
Job retrieved
{(0, 0) : 4, (1, 0) : 6}
```

From a usage perspective, something like the following would be sufficient:

```
.. jupyter-execute::
    :dont-run:
    MyExpensiveCall()

.. jupyter-execute::
    :output:
    Connecting... done
    Job submitted
    Job retrieved
    {(0, 0) : 4, (1, 0) : 6}
```

**Comment by akhmerov** [comment]
OK, I understand what we aim to achieve here. I also imagine that we'd want the inputs and the outputs that bypassed execution would be included in binders/downloaded code/thebe.

Regarding the implementation: we already have ast nodes that correspond to inputs and outputs. Therefore rather than overloading an already somewhat bloated `jupyter-execute` directive with options we could instead expose the nodes that are that are now internal as directives. @willsimmons1465 what do you think?

**Comment by willsimmons1465** [comment]
@akhmerov Sounds fine by me. I have updated the PR to test out adding new directives and adding the input to the downloaded code files. I still need to fix all of the tests that were affected by the changes and tidy the code up a bit, but let me know if this is now on the right tracks.

Given cell outputs are not currently written to the ipynb file output, I also haven't added the manually-specified outputs there. I feel like maybe outputs _should_ be written there but that is probably for a separate discussion

**Comment by caenrigen** [comment]
Hello!

First of all thank you for this nice extension and the developments!! 🙂

I was looking for this kind of functionality for a different use case. Is there any solution already available?
What I am looking for:
1) Skip cell execution (e.g. some directive)
2) Cell is still executable (not raw format) when the `.ipynb` is downloaded by the users

Why?

We are documenting interactions with physical instruments and it is not always possible to have a mock instrument but the notebook needs to still be executable by the users

Looking forward to hear about the developments!
Cheers

**Comment by caenrigen** [comment]
Hello, guys! Hope you are doing well in current times!
Any update on this issue and #141 ? Looking forward to using this feature

**Comment by akhmerov** [comment]
Closed via #141
